pkg/secrets/kubernetes_test.go

```diff
@@ -617,6 +628,114 @@ func TestProvider(t *testing.T) {
 
 		provider.Remove(toSecretConfig(typeBinary.secret, key))
 
+		require.True(t, provider.isClean())
+	})
+	t.Run("network error", func(t *testing.T) {
+		c := fake.NewSimpleClientset(
+			validEmptySecret.DeepCopy(),
+			validBinarySecret.DeepCopy(),
+			validStringSecret.DeepCopy(),
+			validMixedSecret.DeepCopy(),
+		)
+
+		networkError := true
+
+		c.PrependReactor("get", "secrets", func(action clienttesting.Action) (bool, runtime.Object, error) {
+			if networkError {
+				return true, nil, apierrors.NewInternalError(errors.New("network"))
+			}
+			return false, nil, nil
+		})
+
+		ctx := context.Background()
+		provider := newWatchProvider(ctx, log.NewNopLogger(), c)
+
+		key := typeBinary.entries[0].key
+		s, err := provider.Add(toSecretConfig(typeBinary.secret, key))
+		require.NoError(t, err)
+		err = errNotFound(typeBinary.secret.Namespace, typeBinary.secret.Name)
+		requireFetchFail(ctx, t, s, err)
+
+		secret := typeBinary.secret.DeepCopy()
+		updateKey(secret, key, "Goodbye")
+		_, err = c.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+		require.NoError(t, err)
+
+		networkError = false
+		requireFetchEquals(ctx, t, s, "Goodbye")
+
+		// Ensure the new watcher is working.
+		updateKey(secret, key, "Hello")
+		_, err = c.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+		require.NoError(t, err)
+
+		requireFetchEquals(ctx, t, s, "Hello")
```

Comment:
It ensures that the watch provider:
1. Does not error when calling `Add`.
2. Eventually watches the same secret from the initial `Add`.
3. Picks up that the secret changed since the initial `Add` (which comes from the Kubernetes `get` API, as that's the first call that happens after the permission is granted).
4. Picks up that the secret changed again after the last change (which comes from the Kubernetes `watch` API, as `get` is no longer called once a watch is established).

`s` is an interface that implements `Fetch` which stores data from the watch.

Reply:
Awesome - I missed that the `Secret` type of `s` is actually our internal secret, not `corev1.Secret`.

Could you add this as a comment in the test code?

Reply:
Renamed to `providerSecret`.